graders/python/python_autograder/code_feedback.py

```diff
@@ -299,6 +299,103 @@ def bad(msg):
+
+        - ``name``: Name of the dictionary that is being checked. This will be used to give feedback.
+        - ``ref``: Reference dictionary.
+        - ``data``: Student dictionary to be checked.
+        - ``partial_keys``: If not None, it takes a List of keys to check if these particular keys are present in the student's dict or not.
+        - ``check_only_keys``: If true, grading will be done only based on checking all keys in student's dict and reference's dict match or not.
+        - ``check_only_values``: If true, grading will be done only based on checking all valeus in student's dict and reference's dict match or not.
+        - ``entry_type_key``: If not None, requires that each key in the student's dictionary in solution be of this type.
+        - ``entry_type_value``: If not None, requires that each value in the student's dictionary in solution be of this type.
+        - ``accuracy_critical``: If true, grading will halt on failure.
+        - ``report_failure``: If true, feedback will be given on failure.
+        """
+
+        def bad(msg):
+            if report_failure:
+                cls.add_feedback(msg)
+            if accuracy_critical:
+                cls.finish("")
+            else:
+                return False
+
+        if data is None:
+            return bad(f"{name} is None or not defined")
+
+        if not isinstance(data, dict):
+            return bad(f"{name} is not a dict")
+
+        if len(ref) != len(data):
+            return bad(
+                f"{name} has the wrong length for keys--expected {len(ref)}, got {len(data)}"
+            )
+
+        if entry_type_value is not None:
+            for _, value in data.items():
+                if not isinstance(value, entry_type_value):
+                    return bad(f"{name} has the wrong type for value {value}")
+
+        if entry_type_key is not None:
+            for key, _ in data.items():
+                if not isinstance(key, entry_type_key):
+                    return bad(f"{name} has the wrong type for key {key}")
+
+        if partial_keys is not None and len(partial_keys) > 1:
+            for partial_key in partial_keys:
+                if partial_key not in data:
+                    return bad(f"{name} does not contain key {key}")
+            return True
+
+        if check_only_keys or check_only_values:
+            check_keys = False
+            if check_only_keys:
+                for key in data.keys():
+                    if key not in ref.keys():
+                        return bad(f"{name} contains an extra key: {key}")
```

Comment:
No error gets reported if there is a key in ref.keys() that is not in data.keys() (i.e. extra keys). This functionality is confusing since it seems like this use case should be covered by `check_only_keys`, but is instead by `partial_keys`. I think these should be consolidated, or change `partial_keys` into a flag that disables the length equality check.

Reply:
@eliotwrobson thanks for pointing this out! 


The reason why I did not add additional code for "No error gets reported if there is a key in ref.keys() that is not in data.keys() (i.e. extra keys)." functionality is because, we already perform this check as a default check in the check_dict() function Line: 347

Ref: https://github.com/PrairieLearn/PrairieLearn/pull/9436/files#r1574047293

Reply:
Otherwise, I could change the line 347 
`        if len(ref) != len(data):
`
to 

`        if partial_keys != None and len(ref) != len(data):`

So that, when we are only checking for partial keys, we skip the the length equality check.

Reply:
I think that the second choice (adding the partial_keys check) makes things clearer.